calculator_test.go

```diff
@@ -2,23 +2,285 @@ package calculator_test
+	t.Parallel()
+	tests := []BinaryOpTestCase{
+		{in1: 1, in2: 1, want: 1, name: "Divide ones"},
+		{in1: 2, in2: 1, want: 2, name: "Divide 2 by 1"},
+		{in1: 1, in2: 2, want: 0.5, name: "Divide 1 by 2"},
+		{in1: 1, in2: 0, expectsErr: true, name: "Divide positive by zero"},
+		{in1: -1, in2: 0, expectsErr: true, name: "Divide negative by zero"},
+		{in1: 0, in2: 0, expectsErr: true, name: "Divide negative by zero"},
+		{in1: -1, in2: -1, want: 1, name: "Divide negative by zero"},
+		{in1: 2, in2: -1, want: -2, name: "Divide positive by negative"},
+	}
+	for _, testcase := range tests {
+		got, err := calculator.Divide(testcase.in1, testcase.in2)
+		switch {
+		case err != nil && !testcase.expectsErr:
+			t.Errorf("Did not expect an error but got one in %v", testcase.name)
+		case err == nil && testcase.expectsErr:
+			t.Errorf("Expected an error and did NOT get one in %v", testcase.name)
+		case err == nil && !testcase.expectsErr:
+			if testcase.want != got {
+				t.Errorf("Wanted %v, got %v in %v", testcase.want, got, testcase.name)
+			}
+		}
 	}
 }
 
-// func TestSubtract(t *testing.T) {
-// 	t.Parallel()
-// 	var want float64 = 2
-// 	got := calculator.Subtract(4, 2)
-// 	if want != got {
-// 		t.Errorf("want %f, got %f", want, got)
-// 	}
-// }
+func TestDivideMulti(t *testing.T) {
+	testcases := []MultiOpTestCase{
+		{in: []float64{1, 2}, want: .5, name: "Divide 2 numbers"},
+		{in: []float64{1, 2, .5}, want: 1, name: "Divide 3 numbers"},
+		{in: []float64{12, 4, 3}, want: 1, name: "Given example"},
+		{in: []float64{1}, expectsErr: true, name: "Can't divide a single number"},
+	}
+	for _, testcase := range testcases {
+		got, err := calculator.Divide(testcase.in...)
+		switch {
+		case err != nil && !testcase.expectsErr:
+			t.Errorf("Did not expect an error but got one in %v", testcase.name)
+		case err == nil && testcase.expectsErr:
+			t.Errorf("Expected an error and did NOT get one in %v", testcase.name)
+		case err == nil && !testcase.expectsErr:
+			if testcase.want != got {
+				t.Errorf("Wanted %v, got %v in %v", testcase.want, got, testcase.name)
+			}
+
+		}
+	}
+}
+
+func TestSqrt(t *testing.T) {
+	testcases := []UnaryOpTestCase{
+		{in: 4, want: 2, name: "Sqrt 4"},
+		{in: 0, want: 0, name: "Sqrt 0"},
+		{in: 1, want: 1, name: "Sqrt 1"},
+		{in: .25, want: .5, name: "Sqrt .25"},
+		{in: -0.000000001, expectsErr: true, name: "Sqrt small negative number"},
+		{in: -1, expectsErr: true, name: "Sqrt negative number"},
+	}
+	for _, testcase := range testcases {
+		got, err := calculator.Sqrt(testcase.in)
+		switch {
+		case err != nil && !testcase.expectsErr:
+			t.Errorf("Got an error without expecting one for sqrt %v", testcase.in)
+		case err == nil && testcase.expectsErr:
+			t.Errorf("Did not get an error but expected one for sqrt %v", testcase.in)
+		case err == nil && !testcase.expectsErr:
+			if testcase.want != got {
+				t.Errorf("Got %v, wanted %v, using sqrt %v", got, testcase.want, testcase.in)
+			}
+		}
+	}
+}
```

Comment:
I am using this switch case a lot.  Would it be better factored as a function?  Needing to parameterize the error strings seemed like too much complexity.